File: general-solving/leetcode/0226_invert-binary-tree.cs
/***************************************************************************************************
* Title : Invert children of Binary Tree
* URL   : https://leetcode.com/problems/invert-binary-tree/
* Date  : 2015-07-22
* Comp  : O(n)
* Author: Atiq Rahman
* Status: Accepted
* Notes : Simple and beatiful recursion
*   No library Swap function in C#, I have put on in utils source file in this dir.
* rel   : http://www.lintcode.com/en/problem/invert-binary-tree/
*         https://www.hackerrank.com/challenges/swap-nodes-algo
* meta  : tag-binary-tree, tag-recursion, tag-company-microsoft, tag-interview-onsite,
*   tag-leetcode-easy
***************************************************************************************************/
public class Solution
{
  // second version, swap later, still works
  public TreeNode InvertTree(TreeNode root) {
    if (root == null)
      return null;
    InvertTree(root.left);
    InvertTree(root.right);
    Swap<TreeNode>(ref root.left, ref root.right);
    return root;
  }

  // first version, swap first
  public TreeNode InvertTree(TreeNode root) {
    if (root == null)
      return root;
    Swap<TreeNode>(ref root.left, ref root.right);
    InvertTree(root.left);
    InvertTree(root.right);
    return root;
  }
}

File: leetcode/0226_invert-binary-tree.cpp
/***************************************************************************************************
* Title : Find the minimum depth of a binary tree
* URL   : https://leetcode.com/problems/minimum-depth-of-binary-tree/
* Date  : 2015-07-22 (upate)
* Notes : if we don't consider the
*   - case when left or right child is null solution is wrong
*
* Comp  : O(n)
* Author: Atiq Rahman
* Status: Accepted
* Notes : For leet code add returning node
* rel   : http://www.lintcode.com/en/problem/minimum-depth-of-binary-tree/
* meta  : tag-binary-tree, tag-recursion, tag-leetcode-easy
***************************************************************************************************/
class Solution {
public:
  /**
  * @param root: a TreeNode, the root of the binary tree
  * @return: nothing
  */
  void invertBinaryTree(TreeNode *root) {
    // write your code here
    // null node does not be done anything
    if (root == NULL)
      return;

    // has one or both children
    std::swap(root->left, root->right);
    invertBinaryTree(root->left);
    invertBinaryTree(root->right);
  }
};
